examples/wasmer.sh/index.ts

```diff
@@ -0,0 +1,85 @@
+import "xterm/css/xterm.css";
+
+import { SpawnConfig, Tty, Wasmer, init } from "@wasmer/wasix";
+import { Terminal } from "xterm";
+
+const encoder = new TextEncoder();
+
+const packageName = "sharrattj/bash";
+const args: string[] = [];
+const uses: string[] = ["sharrattj/coreutils"];
+
+async function main() {
+    await init();
+
+    const term = new Terminal({ cursorBlink: true, convertEol: true });
+
+    const element = document.getElementById("app")!;
+    term.open(element);
+
+    term.writeln("Starting...");
+    const wasmer = new Wasmer();
+
+    // Attach the TTY
+    const tty = new Tty();
+    tty.state = {...tty.state, cols: term.cols, rows: term.rows};
+    term.onResize(({cols, rows}) => {
+        tty.state = {...tty.state, cols, rows};
+    });
+    wasmer.runtime().set_tty(tty);
+
+    while (true) {
+        await runInstance(term, wasmer, packageName, { args, uses });
+        term.writeln("Rebooting...");
+    }
+}
+
+async function runInstance(term: Terminal, wasmer: Wasmer, packageName: string, config: SpawnConfig) {
+    const instance = await wasmer.spawn(packageName, config);
+    term.clear();
+
+    const stdin: WritableStreamDefaultWriter<Uint8Array> = instance.stdin!.getWriter();
+    term.onData(line => { stdin.write(encoder.encode(line)); });
+
+    const stdout: ReadableStreamDefaultReader<Uint8Array> = instance.stdout.getReader();
+    copyStream(stdout, line => writeMultiline(term, line));
+
+    const stderr: ReadableStreamDefaultReader<Uint8Array> = instance.stderr.getReader();
+    copyStream(stderr, line => writeMultiline(term, line));
+
+    const { code } = await instance.wait();
+
+    if (code != 0) {
+        term.writeln(`\nExit code: ${code}`);
+    }
+}
+
+async function copyStream(reader: ReadableStreamDefaultReader<Uint8Array>, cb: (line: string) => void) {
+    const decoder = new TextDecoder("utf-8");
+
+    while(true) {
+        const {done, value} = await reader.read();
+
+        if (done || !value) {
+            break;
+        }
+        const chunk = decoder.decode(value);
+        cb(chunk);
+    }
+}
```

Comment:
I think this is a bit convoluted.

I think we can do a bit simpler without this method being async (since we never await for copyStream).
We can simply do:

```suggestion
function streamIntoTerm(reader: ReadableStreamDefaultReader<Uint8Array>, term: Terminal) {
  reader.on('data', (chunk) => {
    term.write(chunk);
  });
});
```

Reply:
Yeah, it's pretty convoluted, but it's written that way because `ReadableStreamDefaultReader` isn't an event emitter so `reader.on("data", ...)` isn't available 😞

Reply:
According to MDN, it's possible to use the `for await ... of ...` syntax [because `ReadableStream` is an async iterator](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#async_iteration).

So in theory, we can rewrite `copyStream()` to this:


```ts
async function copyStream(stream: ReadableStream<Uint8Array>, term: Terminal) {
    const decoder = new TextDecoder("utf-8");

    for await (const chunk of stream) {
        term.write(chunk);
    }
}
```

... but Firefox is the only browser that has actually implemented async iterator support for `ReadableStream` so far.

Reply:
Umm... can we operate with the `ReadableStream` instead of the `ReadableStreamDefaultReader`? That would have the event capabilities right? (meaning: `instance.stdout!.getReader()` returns a `ReadableStream`)

Reply:
Oh wait, `instance.stdout!.getReader()` is already a `ReadableStream`... so we should just good to go on changing the typings?

Reply:
[The `ReadableStream` web API](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream) doesn't provide a `on("data", ...)` callback. Maybe you are confusing it with an api from NodeJS?

This is all we get access to in the browser:

![2023-10-20_16-14](https://github.com/wasmerio/wasmer-js/assets/17380079/f5c94345-1472-4be3-a81d-dd8fc23e1f37)

Reply:
Cool, this should work better:

```suggestion
function streamIntoTerm(reader: ReadableStreamDefaultReader<Uint8Array>, term: Terminal) {
  reader.pipeTo(new WritableStream({
    write(chunk) {
      terminal.write(chunk);
    },
  }));
}
```

The main reason I like this is because there aren't any dangling awaits or while(true). It's a bit cleaner